asterius/src/Asterius/GHCi/Internals.hs

```diff
@@ -0,0 +1,415 @@
+      ghciCompiledCoreExprs = IM.empty,
+      ghciLastCompiledCoreExpr = 0,
+      ghciJSSession = Nothing
+    }
+
+asteriusStartIServ :: GHC.HscEnv -> IO GHC.IServ
+asteriusStartIServ hsc_env = do
+  GHC.debugTraceMsg (GHC.hsc_dflags hsc_env) 3 $ GHC.text "asteriusStartIServ"
+  (node_read_fd, host_write_fd) <- createPipeFd
+  (host_read_fd, node_write_fd) <- createPipeFd
+  setEnv "ASTERIUS_NODE_READ_FD" (show node_read_fd) True
+  setEnv "ASTERIUS_NODE_WRITE_FD" (show node_write_fd) True
+  host_read_handle <- fdToHandle host_read_fd
+  host_write_handle <- fdToHandle host_write_fd
+  hSetBuffering host_read_handle NoBuffering
+  hSetBuffering host_write_handle NoBuffering
+  lo_ref <- newIORef Nothing
+  let p = Pipe
+        { pipeRead = host_read_handle,
+          pipeWrite = host_write_handle,
+          pipeLeftovers = lo_ref
+        }
+  s <- newJSSession defJSSessionOpts {nodeStdErrInherit = True}
+  modifyMVar_ globalGHCiState $ \st ->
+    pure
+      st
+        { ghciJSSession =
+            Just
+              (s, p, error "JSVal of asterius instance uninitialized")
+        }
+  cache_ref <- newIORef GHC.emptyUFM
+  pure GHC.IServ
+    { GHC.iservPipe = error "asteriusStartIServ.iservPipe",
+      GHC.iservProcess = error "asteriusStartIServ.iservProcess",
+      GHC.iservLookupSymbolCache = cache_ref,
+      GHC.iservPendingFrees = []
+    }
+
+asteriusStopIServ :: GHC.HscEnv -> IO ()
+asteriusStopIServ hsc_env = do
+  GHC.debugTraceMsg (GHC.hsc_dflags hsc_env) 3 $ GHC.text "asteriusStopIServ"
+  modifyMVar_ globalGHCiState $ \st -> case ghciJSSession st of
+    Just (s, _, _) -> do
+      closeJSSession s
+      pure st {ghciJSSession = Nothing}
+    _ -> pure st
+  pure ()
+
+asteriusIservCall ::
+  Binary a => GHC.HscEnv -> GHC.IServ -> GHC.Message a -> IO a
+asteriusIservCall hsc_env _ msg = do
+  GHC.debugTraceMsg (GHC.hsc_dflags hsc_env) 3 $ GHC.text $ show msg
+  case msg of
+    GHC.InitLinker -> pure ()
+    GHC.LoadDLL _ -> pure Nothing
+    GHC.LoadArchive p -> modifyMVar_ globalGHCiState $ \s -> do
+      lib <- loadAr p
+      evaluate s {ghciLibs = lib <> ghciLibs s}
+    GHC.LoadObj p -> modifyMVar_ globalGHCiState $ \s -> do
+      obj <- decodeFile p
+      evaluate s {ghciObjs = M.insert p obj $ ghciObjs s}
+    GHC.AddLibrarySearchPath _ -> pure $ GHC.RemotePtr 0
+    GHC.RemoveLibrarySearchPath _ -> pure True
+    GHC.ResolveObjs -> pure True
+    GHC.FindSystemLibrary lib -> pure $ Just lib
+    GHC.StartTH -> pure $ unsafeCoerce $ GHC.RemotePtr 0
+    _ -> fail "asteriusIservCall"
+
+asteriusReadIServ ::
+  forall a. (Binary a, Typeable a) => GHC.HscEnv -> GHC.IServ -> IO a
+asteriusReadIServ _ _ = withMVar globalGHCiState $
+  \s -> let Just (_, p, _) = ghciJSSession s in readPipe p get
+
+asteriusWriteIServ ::
+  (Binary a, Typeable a) => GHC.HscEnv -> GHC.IServ -> a -> IO ()
+asteriusWriteIServ hsc_env i a
+  | Just HRefl <- eqTypeRep (typeOf a) (typeRep @GHC.Msg) = case a of
```

Comment:
It would be simpler to replace the explicit type equality check with `case fromDynamic (toDyn a) of ...`